syntax = "proto3";

package slack_action;

// Slack Action Service for receiving tasks from processor
service SlackActionService {
  rpc SendMessage(SendMessageRequest) returns (SendMessageResponse);
  rpc SendEphemeral(SendEphemeralRequest) returns (SendEphemeralResponse);
  rpc UpdateMessage(UpdateMessageRequest) returns (UpdateMessageResponse);
  rpc DeleteMessage(DeleteMessageRequest) returns (DeleteMessageResponse);
  rpc AddReaction(AddReactionRequest) returns (AddReactionResponse);
  rpc RemoveReaction(RemoveReactionRequest) returns (RemoveReactionResponse);
  rpc UploadFile(UploadFileRequest) returns (UploadFileResponse);
  rpc CreateChannel(CreateChannelRequest) returns (CreateChannelResponse);
  rpc InviteToChannel(InviteToChannelRequest) returns (InviteToChannelResponse);
  rpc KickFromChannel(KickFromChannelRequest) returns (KickFromChannelResponse);
  rpc SetTopic(SetTopicRequest) returns (SetTopicResponse);
  rpc OpenModal(OpenModalRequest) returns (OpenModalResponse);
}

// Message requests/responses
message SendMessageRequest {
  string community_id = 1;
  string channel_id = 2;
  string text = 3;
  string blocks_json = 4;  // JSON-encoded blocks
  string thread_ts = 5;    // Optional thread timestamp
}

message SendMessageResponse {
  bool success = 1;
  string message_ts = 2;
  string error = 3;
}

message SendEphemeralRequest {
  string community_id = 1;
  string channel_id = 2;
  string user_id = 3;
  string text = 4;
}

message SendEphemeralResponse {
  bool success = 1;
  string error = 2;
}

message UpdateMessageRequest {
  string community_id = 1;
  string channel_id = 2;
  string ts = 3;
  string text = 4;
  string blocks_json = 5;
}

message UpdateMessageResponse {
  bool success = 1;
  string error = 2;
}

message DeleteMessageRequest {
  string community_id = 1;
  string channel_id = 2;
  string ts = 3;
}

message DeleteMessageResponse {
  bool success = 1;
  string error = 2;
}

// Reaction requests/responses
message AddReactionRequest {
  string community_id = 1;
  string channel_id = 2;
  string ts = 3;
  string emoji = 4;
}

message AddReactionResponse {
  bool success = 1;
  string error = 2;
}

message RemoveReactionRequest {
  string community_id = 1;
  string channel_id = 2;
  string ts = 3;
  string emoji = 4;
}

message RemoveReactionResponse {
  bool success = 1;
  string error = 2;
}

// File upload
message UploadFileRequest {
  string community_id = 1;
  string channel_id = 2;
  bytes file_content = 3;
  string filename = 4;
  string title = 5;
}

message UploadFileResponse {
  bool success = 1;
  string file_id = 2;
  string error = 3;
}

// Channel management
message CreateChannelRequest {
  string community_id = 1;
  string name = 2;
  bool is_private = 3;
}

message CreateChannelResponse {
  bool success = 1;
  string channel_id = 2;
  string error = 3;
}

message InviteToChannelRequest {
  string community_id = 1;
  string channel_id = 2;
  repeated string user_ids = 3;
}

message InviteToChannelResponse {
  bool success = 1;
  string error = 2;
}

message KickFromChannelRequest {
  string community_id = 1;
  string channel_id = 2;
  string user_id = 3;
}

message KickFromChannelResponse {
  bool success = 1;
  string error = 2;
}

message SetTopicRequest {
  string community_id = 1;
  string channel_id = 2;
  string topic = 3;
}

message SetTopicResponse {
  bool success = 1;
  string error = 2;
}

// Modal dialog
message OpenModalRequest {
  string community_id = 1;
  string trigger_id = 2;
  string view_json = 3;  // JSON-encoded view
}

message OpenModalResponse {
  bool success = 1;
  string view_id = 2;
  string error = 3;
}
